lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
+
+        // Step 1: Always include baseConfig
+        configs = [this.baseConfig];
+
+        // Step 2: Add user-specified config from .eslintrc.* and package.json files
         if (this.useEslintrc) {
             debug("Using .eslintrc and package.json files");
-            userConfig = getLocalConfig(this, directory);
+            configs = configs.concat(this.getLocalConfigHierarchy(directory));
         } else {
             debug("Not using .eslintrc or package.json files");
-            userConfig = {};
         }
 
-        // Step 2: Create a copy of the baseConfig
-        config = ConfigOps.merge({}, this.baseConfig);
+        // Step 3: Merge in command line config file
+        if (this.useSpecificConfig) {
+            debug("Using command line config file");
+            configs.push(this.useSpecificConfig);
+        }
 
-        // Step 3: Merge in the user-specified configuration from .eslintrc and package.json
-        config = ConfigOps.merge(config, userConfig);
+        return configs;
+    }
 
-        // Step 4: Merge in command line config file
-        if (this.useSpecificConfig) {
-            debug("Merging command line config file");
+    /**
+     * Gets a list of config objects extracted from local config files that apply to the current directory, in
+     * descending order, beginning with the config that is highest in the directory tree.
+     * @param {string} directory The directory to start looking in for local config files.
+     * @returns {Object[]} The shallow local config objects, or an empty array if there are no local configs.
+     * @private
+     */
+    getLocalConfigHierarchy(directory) {
+        const localConfigFiles = this.findLocalConfigFiles(directory),
+            projectConfigPath = ConfigFile.getFilenameForDirectory(this.options.cwd),
+            searched = [],
+            configs = [];
+        let rootPath,
+            cache;
+
+        for (const localConfigFile of localConfigFiles) {
+            const localConfigDirectory = path.dirname(localConfigFile);
+
+            cache = configCache.getHierarchyLocalConfigs(localConfigDirectory);
+            if (cache) {
+                break;
+            }
 
-            config = ConfigOps.merge(config, this.useSpecificConfig);
-        }
+            // Don't consider the personal config file in the home directory,
+            // except if the home directory is the same as the current working directory
+            if (localConfigDirectory === PERSONAL_CONFIG_DIR && localConfigFile !== projectConfigPath) {
+                continue;
+            }
 
-        // Step 5: Merge in command line environments
-        debug("Merging command line environment settings");
-        config = ConfigOps.merge(config, { env: this.env });
+            // If root flag is set, don't consider file if it is above root
+            if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
```

Comment:
If I'm understanding correctly, this loop will always traverse directories starting with the most nested subdirectory and repeatedly look in the parent directory.

Would it work to just `break;` after a file is found with `root: true`? That way, this `if` statement would be unnecessary because none of the parent directories of the file with `root: true` would get traversed.